README.md
# Follow: CouchDB changes and db updates notifier for NodeJS
[![Build Status](https://travis-ci.org/cloudant-labs/cloudant-follow.svg?branch=master)](https://travis-ci.org/cloudant-labs/cloudant-follow)
[![Greenkeeper badge](https://badges.greenkeeper.io/cloudant-labs/cloudant-follow.svg)](https://greenkeeper.io/)

[![NPM](https://nodei.co/npm/cloudant-follow.png?compact=true)](https://nodei.co/npm/cloudant-follow/)

---

This is a fork of the Iris Couch [follow](https://github.com/iriscouch/follow) project.

This fork is maintained by IBM Cloudant in order to provide fixes for the [nodejs-cloudant](https://github.com/cloudant/nodejs-cloudant) library.

---

## Example

This looks much like the [request][req] API.

```javascript
var follow = require('cloudant-follow');
follow("https://example.iriscouch.com/boogie", function(error, change) {
  if(!error) {
    console.log("Got change number " + change.seq + ": " + change.id);
  }
})
```

The *error* parameter to the callback will basically always be `null`.

## Objective

The API must be very simple: notify me every time a change happens in the DB. Also, never fail.

If an error occurs, Follow will internally retry without notifying your code.

Specifically, this should be possible:

1. Begin a changes feed. Get a couple of change callbacks
2. Shut down CouchDB
3. Go home. Have a nice weekend. Come back on Monday.
4. Start CouchDB with a different IP address
5. Make a couple of changes
6. Update DNS so the domain points to the new IP
7. Once DNS propagates, get a couple more change callbacks

## Failure Mode

If CouchDB permanently crashes, there is an option of failure modes:

* **Default:** Simply never call back with a change again
* **Optional:** Specify an *inactivity* timeout. If no changes happen by the timeout, Follow will signal an error.

## DB Updates

If the db url ends with `/_db_updates`, Follow will provide a
[_db_updates](http://docs.couchdb.org/en/latest/api/server/common.html?highlight=db_updates#get--_db_updates) feed.

For each change, Follow will emit a `change` event containing:

* `type`: `created`, `updated` or `deleted`.
* `db_name`: Name of the database where the change occoured.
* `ok`: Event operation status (boolean).

Note that this feature is available as of CouchDB 1.4.

### Simple API: follow(options, callback)

The first argument is an options object. The only required option is `db`. Instead of an object, you can use a string to indicate the `db` value.

```javascript
follow({db:"https://example.iriscouch.com/boogie", include_docs:true}, function(error, change) {
  if(!error) {
    console.log("Change " + change.seq + " has " + Object.keys(change.doc).length + " fields");
  }
})
```

<a name="options"></a>
All of the CouchDB _changes options are allowed. See http://guide.couchdb.org/draft/notifications.html.

* `db` | Fully-qualified URL of a couch database. (Basic auth URLs are ok.)
* `since` | The sequence number to start from. Use `"now"` to start from the latest change in the DB.
* `heartbeat` | Milliseconds within which CouchDB must respond (default: **30000** or 30 seconds)
* `feed` | Only types "continuous" (or alias "live") _(default)_, "longpoll" and "normal" are supported.
* `filter` |
  * **Either** a path to design document filter, e.g. `app/important`
  * **Or** a Javascript `function(doc, req) { ... }` which should return true or false
* `limit` | Limit the number of changes to the specified value.
* `view`  | A path to design document view, e.g. `app/myView`
* `query_params` | **Optional** for use in with `filter` functions, passed as `req.query` to the filter function

Besides the CouchDB options, more are available:

* `headers` | Object with HTTP headers to add to the request
* `inactivity_ms` | Maximum time to wait between **changes**. Omitting this means no maximum.
* `max_retry_seconds` | Maximum time to wait between retries (default: 360 seconds)
* `initial_retry_delay` | Time to wait before the first retry, in milliseconds (default 1000 milliseconds)
* `response_grace_time` | Extra time to wait before timing out, in milliseconds (default 5000 milliseconds)
* `httpAgent` | Specify a custom HTTP agent to use for all requests (default: `request`).

## Object API

The main API is a thin wrapper around the EventEmitter API.

```javascript
var follow = require('cloudant-follow');

var opts = {}; // Same options paramters as before
var feed = new follow.Feed(opts);

// You can also set values directly.
feed.db            = "http://example.iriscouch.com/boogie";
feed.since         = 3;
feed.heartbeat     = 30    * 1000
feed.inactivity_ms = 86400 * 1000;

feed.filter = function(doc, req) {
  // req.query is the parameters from the _changes request and also feed.query_params.
  console.log('Filtering for query: ' + JSON.stringify(req.query));

  if(doc.stinky || doc.ugly)
    return false;
  return true;
}

feed.on('change', function(change) {
  console.log('Doc ' + change.id + ' in change ' + change.seq + ' is neither stinky nor ugly.');
})

feed.on('error', function(er) {
  console.error('Since Follow always retries on errors, this must be serious');
  throw er;
})

feed.follow();
```

<a name="pause"></a>
## Pause and Resume

A Follow feed is a Node.js stream. If you get lots of changes and processing them takes a while, use `.pause()` and `.resume()` as needed. Pausing guarantees that no new events will fire. Resuming guarantees you'll pick up where you left off.

```javascript
follow("https://example.iriscouch.com/boogie", function(error, change) {
  var feed = this

  if(change.seq == 1) {
    console.log('Uh oh. The first change takes 30 hours to process. Better pause.')
    feed.pause()
    setTimeout(function() { feed.resume() }, 30 * 60 * 60 * 1000)
  }

  // ... 30 hours with no events ...

  else
    console.log('No need to pause for normal change: ' + change.id)
})
```

<a name="events"></a>
## Events

The feed object is an EventEmitter. There are a few ways to get a feed object:

* Use the object API above
* Use the return value of `follow()`
* In the callback to `follow()`, the *this* variable is bound to the feed object.

Once you've got one, you can subscribe to these events:

* **start** | Before any i/o occurs
* **confirm_request** | `function(req)` | The database confirmation request is sent; passed the `request` object
* **confirm** | `function(db_obj)` | The database is confirmed; passed the couch database object
* **change** | `function(change)` | A change occured; passed the change object from CouchDB
* **catchup** | `function(seq_id)` | The feed has caught up to the *update_seq* from the confirm step. Assuming no subsequent changes, you have seen all the data.
* **wait** | Follow is idle, waiting for the next data chunk from CouchDB
* **timeout** | `function(info)` | Follow did not receive a heartbeat from couch in time. The passed object has `.elapsed_ms` set to the elapsed time
* **retry** | `function(info)` | A retry is scheduled (usually after a timeout or disconnection). The passed object has
  * `.since` the current sequence id
  * `.after` the milliseconds to wait before the request occurs (on an exponential fallback schedule)
  * `.db` the database url (scrubbed of basic auth credentials)
* **stop** | The feed is stopping, because of an error, or because you called `feed.stop()`
* **error** | `function(err)` | An error occurs

## Error conditions

Follow is happy to retry over and over, for all eternity. It will only emit an error if it thinks your whole application might be in trouble.

* *DB confirmation* failed: Follow confirms the DB with a preliminary query, which must reply properly.
* *DB is deleted*: Even if it retried, subsequent sequence numbers would be meaningless to your code.
* *Your inactivity timer* expired: This is a last-ditch way to detect possible errors. What if couch is sending heartbeats just fine, but nothing has changed for 24 hours? You know that for your app, 24 hours with no change is impossible. Maybe your filter has a bug? Maybe you queried the wrong DB? Whatever the reason, Follow will emit an error.
* JSON parse error, which should be impossible from CouchDB
* Invalid change object format, which should be impossible from CouchDB
* Internal error, if the internal state seems wrong, e.g. cancelling a timeout that already expired, etc. Follow tries to fail early.

## Tests

Follow uses [node-tap][tap]. If you clone this Git repository, tap is included.

    $ ./node_modules/.bin/tap test/*.js test/issues/*.js
    ok test/couch.js ...................................... 11/11
    ok test/follow.js ..................................... 69/69
    ok test/issues.js ..................................... 44/44
    ok test/stream.js ................................... 300/300
    ok test/issues/10.js .................................. 11/11
    total ............................................... 435/435

    ok

## License

Apache 2.0

[req]: https://github.com/mikeal/request
[tap]: https://github.com/isaacs/node-tap


CHANGES.md
# UNRELEASED
- [NEW] Allow a custom HTTP agent to be specified.
- [NEW] Trim the database URL if it includes the `/_changes` endpoint.
- [NEW] Support type "normal" and type "longpoll" feeds.
- [FIXED] Retry `/_db_updates` request using the last recorded sequence value as
  the `since` query parameter.

# 0.16.1 (2017-11-02)
- [FIXED] Corrected `eslint` to be `devDependencies` instead of `dependencies`.

# 0.16.0 (2017-10-31)
- [UPGRADED] Use latest version of request@2.x.x.

# 0.15.0 (2017-10-30)
- [FIXED] Check `request` is defined before using in feed timeout handler.
- [FIXED] Update list of possible feed query parameters.

# 0.14.0 (2017-09-29)
- [NEW] Correctly construct feed URL for `feed.view`.

# 0.13.0 (2017-05-16)
- [FIXED] Retry changes feed if server sends `last_seq`.


test/params.js
// Copyright © 2017 IBM Corp. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
'use strict';

var tap = require('tap');
var test = tap.test;

var couch = require('./couch');
var follow = require('../api');

couch.setup(test);

test('Supports valid `/_changes` feed parameters', function(t) {
  var feed = new follow.Feed();
  feed.db = couch.DB;

  var params = {};
  params.conflicts = false;
  params.descending = false;
  params.feed = 'continuous';
  params.heartbeat = 12345;
  params.include_docs = false;
  params.limit = 12345;
  params.seq_interval = 1;
  params.since = 0;
  params.style = 'main_only';
  params.timeout = 12345;

  // add parameters
  Object.keys(params).forEach(function(param) {
    feed[param] = params[param];
  });

  feed.on('query', function(req) {
    var query = [];
    Object.keys(params).forEach(function(param) {
      query.push(`${param}=${params[param]}`);
    });

    t.equal(req.uri.query, query.join('&'));
    t.end();
  });

  feed.on('change', function() {
    this.stop();
  });

  feed.follow();
});

test('Ignore invalid `/_changes` feed parameters', function(t) {
  var feed = new follow.Feed();
  feed.db = couch.DB;

  feed.foo = 'bar'; // invalid feed parameter

  feed.on('query', function(req) {
    t.equal(req.uri.query, 'feed=continuous&heartbeat=30000&seq_interval=1&since=0');
    t.end();
  });

  feed.on('change', function() {
    this.stop();
  });

  feed.follow();
});


lib/feed.js
// Copyright © 2017 IBM Corp. All rights reserved.
// Copyright © 2011 Jason Smith, Jarrett Cruger and contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
'use strict';

// Core routines for event emitters

var lib = require('../lib');
var util = require('util');
var events = require('events');
var request = require('request');
var Changes = require('./stream').Changes;

// Use the library timeout functions, primarily so the test suite can catch errors.
var setTimeout = lib.setTimeout;
var clearTimeout = lib.clearTimeout;

var DEFAULT_HEARTBEAT = 30000;
var HEARTBEAT_TIMEOUT_COEFFICIENT = 1.25; // E.g. heartbeat 1000ms would trigger a timeout after 1250ms of no heartbeat.
var DEFAULT_MAX_RETRY_SECONDS = 60 * 60;
var INITIAL_RETRY_DELAY = 1000;
var RESPONSE_GRACE_TIME = 5000;
var DEFAULT_SEQ_INTERVAL = 1;

// https://console.bluemix.net/docs/services/Cloudant/api/database.html#get-changes
var CHANGES_PARAMETERS = [
  'conflicts',
  'descending',
  'doc_ids',
  'feed',
  'filter',
  'heartbeat',
  'include_docs',
  'limit',
  'seq_interval',
  'since',
  'style',
  'timeout',

  // additional parameters as defined in http://docs.couchdb.org/en/2.0.0/api/database/changes.html
  'att_encoding_info',
  'attachments',
  'last-event-id',
  'view'
];

// https://console.bluemix.net/docs/services/Cloudant/api/advanced.html#-get-_db_updates-
var DB_UPDATES_PARAMETERS = [
  'descending',
  'feed',
  'heartbeat',
  'limit',
  'since',
  'timeout'
];

var EventEmitter = events.EventEmitter2 || events.EventEmitter;

util.inherits(Feed, EventEmitter);
function Feed(opts) {
  var self = this;
  EventEmitter.call(self);

  opts = opts || {};

  self.feed = opts.feed || 'continuous';
  self.heartbeat = opts.heartbeat || DEFAULT_HEARTBEAT;
  self.max_retry_seconds = opts.max_retry_seconds || DEFAULT_MAX_RETRY_SECONDS;
  self.inactivity_ms = null;
  self.initial_retry_delay = opts.initial_retry_delay || INITIAL_RETRY_DELAY;
  self.response_grace_time = opts.response_grace_time || RESPONSE_GRACE_TIME;
  self.seq_interval = opts.seq_interval || DEFAULT_SEQ_INTERVAL;

  if (typeof opts.httpAgent === 'function') {
    self.httpAgent = opts.httpAgent;
  } else {
    self.httpAgent = request.defaults(opts.httpAgent || {});
  }

  self.headers = {};
  self.request = opts.request || {}; // Extra options for potentially future versions of request. The caller can supply them.

  self.since = 0;
  self.is_paused = false;
  self.caught_up = false;
  self.seen_last_seq = false;
  self.retry_delay = self.initial_retry_delay;

  self.query_params = {}; // Extra `req.query` values for filter functions

  if (typeof opts === 'string') {
    opts = {'db': opts};
  }

  Object.keys(opts).forEach(function(key) {
    if (typeof self[key] !== 'function') {
      self[key] = opts[key];
    }
  });

  self.pending = { request: null, activity_at: null };
} // Feed

Feed.prototype.start =
Feed.prototype.follow = function follow_feed() {
  var self = this;

  self.db = self.db || self.url || self.uri;
  delete self.url;
  delete self.uri;

  if (!self.db) {
    throw new Error('Database URL required');
  }

  if ([ 'continuous', 'live', 'longpoll', 'normal' ].indexOf(self.feed) === -1) {
    throw new Error(`Invalid feed option: ${self.feed}`);
  }

  if (self.feed === 'live') {
    self.feed = 'continuous'; // alias
  }

  if (self.db.match(/\/_db_updates$/)) {
    // using `/_db_updates`
    self.is_db_updates = true;
  } else if (self.db.match(/\/_changes$/)) {
    // using `/<db>/_changes`
    self.db = self.db.replace(/\/_changes$/, ''); // remove unnecessary endpoint
  }

  if (typeof self.heartbeat !== 'number') {
    throw new Error('Required "heartbeat" value');
  }

  self.log = lib.log4js.getLogger(self.db);
  self.log.setLevel(process.env.follow_log_level || 'info');

  self.emit('start');
  return self.confirm();
};

Feed.prototype.confirm = function confirm_feed() {
  var self = this;

  self.db_safe = lib.scrub_creds(self.db);

  var endpoint = self.is_db_updates ? 'server' : 'database';

  self.log.debug('Checking ' + endpoint + ': ' + self.db_safe);

  var confirm_timeout = self.heartbeat * 3; // Give it time to look up the name, connect, etc.
  var timeout_id = setTimeout(function() {
    return self.die(new Error('Timeout confirming ' + endpoint + ': ' + self.db_safe));
  }, confirm_timeout);

  var headers = lib.JP(lib.JS(self.headers));
  headers.accept = 'application/json';

  var uri = self.is_db_updates ? self.db.replace(/\/_db_updates$/, '') : self.db;
  var req = {'uri': uri, 'headers': headers};
  Object.keys(self.request).forEach(function(key) {
    req[key] = self.request[key];
  });

  req = self.httpAgent(req, db_response);
  self.emit('confirm_request', req);

  function db_response(er, resp, body) {
    clearTimeout(timeout_id);

    if (er) {
      return self.die(er);
    }

    var db;
    try {
      db = JSON.parse(body);
    } catch (json_er) {
      return self.emit('error', json_er);
    }

    if (!self.is_db_updates && !self.dead && (!db.db_name || !db.instance_start_time)) {
      return self.emit('error', new Error('Bad DB response: ' + body));
    }

    if (self.is_db_updates && !self.dead && !db.couchdb) {
      return self.emit('error', new Error('Bad server response: ' + body));
    }

    if (!self.is_db_updates) {
      self.original_db_seq = db.update_seq;
    }

    self.log.debug('Confirmed ' + endpoint + ': ' + self.db_safe);
    self.emit('confirm', db);

    if (self.since === 'now') {
      self.log.debug('Query since "now" is the same as query since -1');
      self.since = -1;
    }

    if (self.since === -1) {
      self.log.debug('Query since ' + self.since + ' will start at ' + db.update_seq);
      self.since = db.update_seq;
    } else if (self.since < 0) {
      if (isNaN(db.update_seq)) {
        return self.emit('error', new Error('DB requires specific id in "since"'));
      }

      self.log.debug('Query since ' + self.since + ' will start at ' + (db.update_seq + self.since + 1));
      self.since = db.update_seq + self.since + 1;
    }

    // If the next change would come after the current update_seq, just fake a catchup event now.
    if (self.original_db_seq === self.since) {
      self.caught_up = true;
      self.emit('catchup', db.update_seq);
    }

    return self.query();
  }
};

Feed.prototype.query = function query_feed() {
  var self = this;

  var args;
  if (self.is_db_updates) {
    args = DB_UPDATES_PARAMETERS;
  } else {
    args = CHANGES_PARAMETERS;
  }

  var query_params = JSON.parse(JSON.stringify(self.query_params));

  args.forEach(function(key) {
    if (key in self) {
      query_params[key] = self[key];
    }
  });

  if (typeof query_params.filter !== 'string') {
    delete query_params.filter;
  }

  if (typeof self.filter === 'function' && !query_params.include_docs) {
    self.log.debug('Enabling include_docs for client-side filter');
    query_params.include_docs = true;
  }

  if (query_params.view) {
    query_params.filter = '_view';
  }

  self.query_params = query_params;

  var feed_url = self.db + (self.is_db_updates ? '' : '/_changes');

  self.headers.accept = self.headers.accept || 'application/json';
  var req = {
    method: 'GET',
    uri: feed_url,
    qs: query_params,
    headers: self.headers,
    encoding: 'utf-8'
  };

  req.changes_query = query_params;
  Object.keys(self.request).forEach(function(key) {
    req[key] = self.request[key];
  });

  var now = new Date();
  var feed_ts = lib.JDUP(now);
  var feed_id = process.env.follow_debug ? feed_ts.match(/\.(\d\d\d)Z$/)[1] : feed_ts;

  self.log.debug('Feed query ' + feed_id + ': ' + lib.scrub_creds(feed_url));
  var feed_request = self.httpAgent(req);

  feed_request.on('response', function(res) {
    self.log.debug('Remove feed from agent pool: ' + feed_id);
    res.req.socket.emit('agentRemove');

    // Simulate the old onResponse option.
    on_feed_response(null, res, res.body);
  });

  feed_request.on('error', on_feed_response);

  // The response headers must arrive within one heartbeat.
  var response_timer = setTimeout(response_timed_out, self.heartbeat + self.response_grace_time);
  var timed_out = false;

  return self.emit('query', feed_request);

  function response_timed_out() {
    self.log.debug('Feed response timed out: ' + feed_id);
    timed_out = true;
    return self.retry();
  }

  function on_feed_response(er, resp, body) {
    clearTimeout(response_timer);

    if ((resp !== undefined && resp.body) || body) {
      return self.die(new Error('Cannot handle a body in the feed response: ' + lib.JS(resp.body || body)));
    }

    if (timed_out) {
      self.log.debug('Ignoring late response: ' + feed_id);
      return destroy_response(resp);
    }

    if (er) {
      self.log.debug('Request error ' + feed_id + ': ' + er.stack);
      destroy_response(resp);
      return self.retry();
    }

    if (resp.statusCode === 404) {
      destroy_response(resp);
      self.log.warn('Database not found. Stopping changes feed after seq ' + self.since);
      var del_er = new Error('Database deleted after change: ' + self.since);
      del_er.deleted = true;
      del_er.last_seq = self.since;
      return self.die(del_er);
    }

    if (resp.statusCode !== 200) {
      self.log.debug('Bad changes response ' + feed_id + ': ' + resp.statusCode);
      destroy_response(resp);
      return self.retry();
    }

    self.log.debug('Good response: ' + feed_id);
    self.retry_delay = self.initial_retry_delay;

    self.emit('response', resp);

    var changes_stream = new Changes();
    changes_stream.log = lib.log4js.getLogger('stream ' + self.db);
    changes_stream.log.setLevel(self.log.level.levelStr);
    changes_stream.feed = self.feed;
    resp.pipe(changes_stream);

    changes_stream.created_at = now;
    changes_stream.id = function() {
      return feed_id;
    };
    return self.prep(changes_stream);
  }
};

Feed.prototype.prep = function prep_request(changes_stream) {
  var self = this;

  var now = new Date();
  self.pending.request = changes_stream;
  self.pending.activity_at = now;
  self.pending.wait_timer = null;

  // Just re-run the pause or resume to do the needful on changes_stream (self.pending.request).
  if (self.is_paused) {
    self.pause();
  } else {
    self.resume();
  }

  // The inactivity timer is for time between *changes*, or time between the
  // initial connection and the first change. Therefore it goes here.
  self.change_at = now;
  if (self.inactivity_ms) {
    clearTimeout(self.inactivity_timer);
    self.inactivity_timer = setTimeout(function() {
      self.on_inactivity();
    }, self.inactivity_ms);
  }

  changes_stream.on('heartbeat', handler_for('heartbeat'));
  changes_stream.on('error', handler_for('error'));
  changes_stream.on('data', handler_for('data'));
  changes_stream.on('end', handler_for('end'));

  return self.wait();

  function handler_for(ev) {
    var name = 'on_couch_' + ev;
    var inner_handler = self[name];

    return handle_confirmed_req_event;
    function handle_confirmed_req_event() {
      if (self.pending.request === changes_stream) {
        return inner_handler.apply(self, arguments);
      }

      if (!changes_stream.created_at) {
        return self.die(new Error('Received data from unknown request')); // Pretty sure this is impossible.
      }

      var s_to_now = (new Date() - changes_stream.created_at) / 1000;
      var s_to_req = '[no req]';
      if (self.pending.request) {
        s_to_req = (self.pending.request.created_at - changes_stream.created_at) / 1000;
      }

      var msg = ': ' + changes_stream.id() + ' to_req=' + s_to_req + 's, to_now=' + s_to_now + 's';

      if (ev === 'end' || ev === 'data' || ev === 'heartbeat') {
        self.log.debug('Old "' + ev + '": ' + changes_stream.id());
        return destroy_req(changes_stream);
      }

      self.log.warn('Old "' + ev + '"' + msg);
    }
  }
};

Feed.prototype.wait = function wait_for_event() {
  var self = this;
  self.emit('wait');

  if (self.pending.wait_timer) {
    return self.die(new Error('wait() called but there is already a wait_timer: ' + self.pending.wait_timer));
  }

  var timeout_ms = self.heartbeat * HEARTBEAT_TIMEOUT_COEFFICIENT;
  var req_id = self.pending.request && self.pending.request.id();
  var msg = 'Req ' + req_id + ' timeout=' + timeout_ms;
  if (self.inactivity_ms) {
    msg += ', inactivity=' + self.inactivity_ms;
  }
  msg += ': ' + self.db_safe;

  self.log.debug(msg);
  self.pending.wait_timer = setTimeout(function() {
    self.on_timeout();
  }, timeout_ms);
};

Feed.prototype.got_activity = function() {
  var self = this;

  if (self.dead) {
    return;
  }

  //
  // We may not have a wait_timer so just clear it and null it out if it does
  // exist
  //
  clearTimeout(self.pending.wait_timer);
  self.pending.wait_timer = null;
  self.pending.activity_at = new Date();
};

Feed.prototype.pause = function() {
  var self = this;
  var was_paused = self.is_paused;

  // Emit pause after pausing the stream, to allow listeners to react.
  self.is_paused = true;
  if (self.pending && self.pending.request && self.pending.request.pause) {
    self.pending.request.pause();
  } else {
    self.log.warn('No pending request to pause');
  }

  if (!was_paused) {
    self.emit('pause');
  }
};

Feed.prototype.resume = function() {
  var self = this;
  var was_paused = self.is_paused;

  // Emit resume before resuming the data feed, to allow listeners to prepare.
  self.is_paused = false;
  if (was_paused) {
    self.emit('resume');
  }

  if (self.pending && self.pending.request && self.pending.request.resume) {
    self.pending.request.resume();
  } else {
    self.log.warn('No pending request to resume');
  }
};

Feed.prototype.on_couch_heartbeat = function on_couch_heartbeat() {
  var self = this;

  self.got_activity();
  if (self.dead) {
    return self.log.debug('Skip heartbeat processing for dead feed');
  }

  self.emit('heartbeat');

  if (self.dead) {
    return self.log.debug('No wait: heartbeat listener stopped this feed');
  }
  self.wait();
};

Feed.prototype.on_couch_data = function on_couch_data(change) {
  var self = this;
  self.log.debug('Data from ' + self.pending.request.id());

  self.got_activity();
  if (self.dead) {
    return self.log.debug('Skip data processing for dead feed');
  }

  // The changes stream guarantees that this data is valid JSON.
  change = JSON.parse(change);

  // self.log.debug('Object:\n' + util.inspect(change));
  if ('last_seq' in change) {
    var stopFeed = false;
    if (change.pending === undefined) {
      stopFeed = true;
    } else if (change.pending === 0) {
      stopFeed = true;
    } else if (self.query_params.limit > 0) {
      stopFeed = true;
    }

    if (stopFeed) {
      self.seen_last_seq = true;
      self.emit('last_seq', change.last_seq);
      return self.stop();
    } else {
      self.log.debug('Found last_seq in change. Retrying from seq ' + change.last_seq);
      self.since = change.last_seq;
      return self.retry();
    }
  }

  if (!self.is_db_updates && self.seq_interval === 1 && !change.seq) {
    return self.die(new Error('Change has no .seq field: ' + JSON.stringify(change)));
  }

  self.on_change(change);

  // on_change() might work its way all the way to a "change" event, and the listener
  // might call .stop(), which means among other things that no more events are desired.
  // The die() code sets a self.dead flag to indicate this.
  if (self.dead) {
    return self.log.debug('No wait: change listener stopped this feed');
  }
  self.wait();
};

Feed.prototype.on_timeout = function on_timeout() {
  var self = this;
  if (self.dead) {
    return self.log.debug('No timeout: change listener stopped this feed');
  }

  self.log.debug('Timeout');

  var now = new Date();
  var elapsed_ms = now - self.pending.activity_at;

  var data = { elapsed_ms: elapsed_ms, heartbeat: self.heartbeat };
  if (self.pending.request) {
    data.id = self.pending.request.id();
  } else {
    self.log.debug('Timeout called for dead feed');
  }

  self.emit('timeout', data);

  /*
  var msg = ' for timeout after ' + elapsed_ms + 'ms; heartbeat=' + self.heartbeat;
  if(!self.pending.request.id)
    self.log.warn('Closing req (no id) ' + msg + ' req=' + util.inspect(self.pending.request));
  else
    self.log.warn('Closing req ' + self.pending.request.id() + msg);
  */

  destroy_req(self.pending.request);
  self.retry();
};

Feed.prototype.retry = function retry() {
  var self = this;

  clearTimeout(self.pending.wait_timer);
  self.pending.wait_timer = null;

  self.log.debug('Retry since=' + self.since + ' after ' + self.retry_delay + 'ms ');
  self.emit('retry', {since: self.since, after: self.retry_delay, db: self.db_safe});

  self.retry_timer = setTimeout(function() {
    self.query();
  }, self.retry_delay);

  var max_retry_ms = self.max_retry_seconds * 1000;
  self.retry_delay *= 2;
  if (self.retry_delay > max_retry_ms) {
    self.retry_delay = max_retry_ms;
  }
};

Feed.prototype.on_couch_end = function on_couch_end() {
  var self = this;

  self.log.debug('Changes feed ended ' + self.pending.request.id());
  self.pending.request = null;

  if (self.seen_last_seq) {
    return self.stop();
  }
};

Feed.prototype.on_couch_error = function on_couch_error(er) {
  var self = this;

  self.log.debug('Changes query eror: ' + lib.JS(er.stack));
  return self.retry();
};

Feed.prototype.stop = function(val) {
  var self = this;
  self.log.debug('Stop');

  // Die with no errors.
  self.die();
  self.emit('stop', val);
};

Feed.prototype.die = function(er) {
  var self = this;

  if (er) {
    self.log.fatal('Fatal error: ' + er.stack);
  }

  // Warn code executing later that death has occured.
  self.dead = true;

  clearTimeout(self.retry_timer);
  clearTimeout(self.inactivity_timer);
  clearTimeout(self.pending.wait_timer);

  self.inactivity_timer = null;
  self.pending.wait_timer = null;

  var req = self.pending.request;
  self.pending.request = null;
  if (req) {
    self.log.debug('Destroying req ' + req.id());
    destroy_req(req);
  }

  if (er) {
    self.emit('error', er);
  }
};

Feed.prototype.on_change = function on_change(change) {
  var self = this;

  if (!self.is_db_updates && self.seq_interval === 1 && !change.seq) {
    return self.die(new Error('No seq value in change: ' + lib.JS(change)));
  }

  if (!self.is_db_updates && self.seq_interval === 1 && change.seq === self.since) {
    self.log.debug('Bad seq value ' + change.seq + ' since=' + self.since);
    return destroy_req(self.pending.request);
  }

  if (typeof self.filter !== 'function') {
    return self.on_good_change(change);
  }

  var req = lib.JDUP({'query': self.pending.request.changes_query});
  var filter_args;

  if (self.is_db_updates) {
    if (!change.db_name || !change.type) {
      return self.die(new Error('Internal _db_updates filter needs .db_name and .type in change ', change));
    }
    filter_args = [change.db_name, change.type, req];
  } else {
    if (!change.doc) {
      return self.die(new Error('Internal filter needs .doc in change ' + change.seq));
    }

    // Don't let the filter mutate the real data.
    var doc = lib.JDUP(change.doc);
    filter_args = [doc, req];
  }
  var result = false;
  try {
    result = self.filter.apply(null, filter_args);
  } catch (er) {
    self.log.debug('Filter error', er);
  }

  result = (result && true) || false;
  if (result) {
    self.log.debug('Builtin filter PASS for change: ' + change.seq);
    return self.on_good_change(change);
  } else {
    self.log.debug('Builtin filter FAIL for change: ' + change.seq);

    // Even with a filtered change, a "catchup" event might still be appropriate.
    self.check_for_catchup(change.seq);
  }
};

Feed.prototype.on_good_change = function on_good_change(change) {
  var self = this;

  if (self.inactivity_ms && !self.inactivity_timer) {
    return self.die(new Error('Cannot find inactivity timer during change'));
  }

  clearTimeout(self.inactivity_timer);
  self.inactivity_timer = null;
  if (self.inactivity_ms) {
    self.inactivity_timer = setTimeout(function() {
      self.on_inactivity();
    }, self.inactivity_ms);
  }

  self.change_at = new Date();

  if (change && change.seq) {
    self.since = change.seq;
  }

  self.emit('change', change);

  self.check_for_catchup(change.seq);
};

Feed.prototype.check_for_catchup = function check_for_catchup(seq) {
  var self = this;

  if (self.is_db_updates) { return; }
  if (self.caught_up) { return; }
  if (seq < self.original_db_seq) { return; }

  self.caught_up = true;
  self.emit('catchup', seq);
};

Feed.prototype.on_inactivity = function on_inactivity() {
  var self = this;
  var now = new Date();
  var elapsed_ms = now - self.change_at;
  var elapsed_s = elapsed_ms / 1000;

  //
  // Since this is actually not fatal, lets just totally reset and start a new
  // request, JUST in case something was bad.
  //
  self.log.debug('Req ' + self.pending.request.id() + ' made no changes for ' + elapsed_s + 's');
  return self.restart();
};

Feed.prototype.restart = function restart() {
  var self = this;

  self.emit('restart');

  // Kill ourselves and then start up once again
  self.stop();
  self.dead = false;
  self.start();
};

module.exports = { 'Feed': Feed };

/*
 * Utilities
 */

function destroy_req(req) {
  if (req) {
    destroy_response(req.response);
  }

  if (req && typeof req.destroy === 'function') {
    req.destroy();
  }
}

function destroy_response(response) {
  if (!response) {
    return;
  }

  if (typeof response.abort === 'function') {
    response.abort();
  }

  if (response.connection) {
    response.connection.destroy();
  }
}
